src/main/java/io/debezium/connector/vitess/pipeline/txmetadata/ShardLineage.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.debezium.connector.vitess.pipeline.txmetadata;
+
+import java.util.Map;
+
+/**
+ * Class used to determine which parents a shard range descended from. Used to set the epoch to the succeediing
+ * epoch of its parents.
+ */
+public class ShardLineage {
+
+    /**
+     * Return the epoch value of the shard, based on its parents epochs.
+     * If there are parents present, return the max of the parent epochs plus one.
+     * If there are no parents present, it returns zero.
+     *
+     * @param shardString The descendant shard to find parents of
+     * @param shardEpochMap The map to search for parents
+     * @return The epoch value of the descendant shard
+     */
+    public static Long getInheritedEpoch(String shardString, ShardEpochMap shardEpochMap) {
+        Shard shard = new Shard(shardString);
+
+        Long maxParentEpoch = -1L;
+        for (Map.Entry<String, Long> shardEpoch : shardEpochMap.getMap().entrySet()) {
+            String currentShardString = shardEpoch.getKey();
+            Long currentEpoch = shardEpoch.getValue();
+            Shard currentShard = new Shard(currentShardString);
+            if (shard.overlaps(currentShard)) {
+                maxParentEpoch = Math.max(maxParentEpoch, currentEpoch);
+            }
+        }
+
+        return maxParentEpoch + 1;
+    }
+
+    private static class Shard {
+
+        // A string lexicographically less than all other strings
+        public static final String NEGATIVE_INFINITY = "";
+        // A string lexicographically greater than all other strings
+        public static final String POSITIVE_INFINITY = "\uFFFF";
+
+        private final String lowerBound;
+        private final String upperBound;
+
+        Shard(String shard) {
+            String[] shardInterval = getShardInterval(shard.toLowerCase());
+            this.lowerBound = getLowerBound(shardInterval);
+            this.upperBound = getUpperBound(shardInterval);
+            validateBounds();
+        }
+
+        private void validateBounds() {
+            if (this.lowerBound.compareTo(this.upperBound) >= 0) {
+                throw new IllegalArgumentException("Invalid shard range " + this);
+            }
+        }
+
+        public boolean overlaps(Shard shard) {
+            return this.lowerBound.compareTo(shard.upperBound) < 0 && this.upperBound.compareTo(shard.lowerBound) > 0;
```

Comment:
How about the situation when the two ranges meet on the boundary ( == 0 case)?

Reply:
Vitess ranges are have inclusive lower bounds and exclusive upper bounds so this should be correct ([docs](https://vitess.io/docs/19.0/reference/features/sharding/#shard-names)). We can think about each part:
1. `this.lowerBound.compareTo(shard.upperBound) == 0` the lower bound is the same as the upper bound we are comparing to. Since the upper bound is exclusive, this interval does not overlap.
2. `this.upperBound.compareTo(shard.lowerBound) == 0` the upper bound is the same as the lower bound we are comparing to. Since the upper bound is exclusive, these two intervals do not overlap.

I added a test to demonstrate this case.